File: mkdocs/snippets/devbook/namespaces/get_namespace_info.mjs
const NODE_URL = process.env.NODE_URL ||
	'http://libertalia.nemtest.net:7890';
console.log('Using node', NODE_URL);

const NAMESPACE_NAME = process.env.NAMESPACE_NAME || 'company';
console.log('Namespace name:', NAMESPACE_NAME);

try {
	// Fetch namespace information [>step-1]
	const namespacePath = `/namespace?namespace=${NAMESPACE_NAME}`;
	console.log('Fetching namespace information from', namespacePath);
	const namespaceResponse = await fetch(`${NODE_URL}${namespacePath}`);
	if (!namespaceResponse.ok)
		throw new Error(`HTTP error! status: ${namespaceResponse.status}`);

	const namespaceInfo = await namespaceResponse.json();
	console.log('Namespace information:');
	console.log('  Name:', namespaceInfo.fqn);
	console.log('  Owner:', namespaceInfo.owner);
	const leaseHeight = namespaceInfo.height;
	console.log('  Height:', leaseHeight);
	// [<step-1]
	// Compute the lease expiration [>step-2]
	const LEASE_DURATION = 525600; // approximately one year of blocks
	const chainResponse = await fetch(`${NODE_URL}/chain/height`);
	const currentHeight = (await chainResponse.json()).height;
	const expirationHeight = leaseHeight + LEASE_DURATION;
	console.log('\nCurrent chain height:', currentHeight);
	console.log('Lease expiration height:', expirationHeight);
	console.log('Blocks until expiration:',
		expirationHeight - currentHeight);
	// [<step-2]
	// List the subnamespaces [>step-3]
	const owner = namespaceInfo.owner;
	const subnamespacesPath = '/account/namespace/page' +
		`?address=${owner}&parent=${NAMESPACE_NAME}`;
	console.log('\nFetching subnamespaces from', subnamespacesPath);
	const subnamespacesResponse =
		await fetch(`${NODE_URL}${subnamespacesPath}`);
	const subnamespaces = (await subnamespacesResponse.json()).data;
	console.log(`Subnamespaces of ${NAMESPACE_NAME}:`,
		subnamespaces.length);
	for (const subnamespace of subnamespaces)
		console.log(`  ${subnamespace.fqn}`);
	// [<step-3]
	// List the mosaics defined under the namespace [>step-4]
	const mosaicsPath =
		`/namespace/mosaic/definition/page?namespace=${NAMESPACE_NAME}`;
	console.log('\nFetching mosaic definitions from', mosaicsPath);
	const mosaicsResponse = await fetch(`${NODE_URL}${mosaicsPath}`);
	const mosaics = (await mosaicsResponse.json()).data;
	console.log(`Mosaics defined under ${NAMESPACE_NAME}:`,
		mosaics.length);
	for (const entry of mosaics) {
		const mosaicId = entry.mosaic.id;
		console.log(`  ${mosaicId.namespaceId}:${mosaicId.name}`);
	}
	// [<step-4]
} catch (e) {
	console.error(e.message);
}

File: mkdocs/snippets/devbook/namespaces/GetNamespaceInfo.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.3

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public final class GetNamespaceInfo {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private static final String NODE_URL = System.getenv().getOrDefault(
		"NODE_URL", "http://libertalia.nemtest.net:7890");

	public static void main(final String[] args) {
		try {
			new GetNamespaceInfo().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", NODE_URL);

		final String namespaceName = System.getenv().getOrDefault(
			"NAMESPACE_NAME", "company");
		System.out.printf("Namespace name: %s%n", namespaceName);

		// Fetch namespace information [>step-1]
		final String namespacePath = String.format(
			"/namespace?namespace=%s", namespaceName);
		System.out.printf("Fetching namespace information from %s%n",
			namespacePath);
		final HttpRequest namespaceRequest = HttpRequest.newBuilder(
			URI.create(NODE_URL + namespacePath)).GET().build();
		final HttpResponse<String> namespaceResponse = HTTP_CLIENT.send(
			namespaceRequest, BodyHandlers.ofString());
		if (2 != namespaceResponse.statusCode() / 100)
			throw new IOException("HTTP error! status: "
				+ namespaceResponse.statusCode());
		final JsonNode namespaceInfo = JSON_MAPPER.readTree(
			namespaceResponse.body());
		System.out.println("Namespace information:");
		System.out.printf("  Name: %s%n",
			namespaceInfo.get("fqn").asText());
		System.out.printf("  Owner: %s%n",
			namespaceInfo.get("owner").asText());
		final long leaseHeight = namespaceInfo.get("height").asLong();
		System.out.printf("  Height: %d%n", leaseHeight);
		// [<step-1]
		// Compute the lease expiration [>step-2]
		final long leaseDuration = 525_600;
		final HttpRequest heightRequest = HttpRequest.newBuilder(
			URI.create(NODE_URL + "/chain/height")).GET().build();
		final HttpResponse<String> heightResponse = HTTP_CLIENT.send(
			heightRequest, BodyHandlers.ofString());
		final long currentHeight = JSON_MAPPER.readTree(
			heightResponse.body()).get("height").asLong();
		final long expirationHeight = leaseHeight + leaseDuration;
		System.out.printf("%nCurrent chain height: %d%n", currentHeight);
		System.out.printf("Lease expiration height: %d%n",
			expirationHeight);
		System.out.printf("Blocks until expiration: %d%n",
			expirationHeight - currentHeight);
		// [<step-2]
		// List the subnamespaces [>step-3]
		final String owner = namespaceInfo.get("owner").asText();
		final String subnamespacesPath = String.format(
			"/account/namespace/page?address=%s&parent=%s",
			owner, namespaceName);
		System.out.printf("%nFetching subnamespaces from %s%n",
			subnamespacesPath);
		final HttpRequest subnamespacesRequest = HttpRequest.newBuilder(
			URI.create(NODE_URL + subnamespacesPath)).GET().build();
		final HttpResponse<String> subnamespacesResponse =
			HTTP_CLIENT.send(
				subnamespacesRequest, BodyHandlers.ofString());
		final JsonNode subnamespaces = JSON_MAPPER.readTree(
			subnamespacesResponse.body()).get("data");
		System.out.printf("Subnamespaces of %s: %d%n",
			namespaceName, subnamespaces.size());
		for (final JsonNode subnamespace : subnamespaces)
			System.out.printf("  %s%n",
				subnamespace.get("fqn").asText());
		// [<step-3]
		// List the mosaics defined under the namespace [>step-4]
		final String mosaicsPath = String.format(
			"/namespace/mosaic/definition/page?namespace=%s",
			namespaceName);
		System.out.printf("%nFetching mosaic definitions from %s%n",
			mosaicsPath);
		final HttpRequest mosaicsRequest = HttpRequest.newBuilder(
			URI.create(NODE_URL + mosaicsPath)).GET().build();
		final HttpResponse<String> mosaicsResponse = HTTP_CLIENT.send(
			mosaicsRequest, BodyHandlers.ofString());
		final JsonNode mosaics = JSON_MAPPER.readTree(
			mosaicsResponse.body()).get("data");
		System.out.printf("Mosaics defined under %s: %d%n",
			namespaceName, mosaics.size());
		for (final JsonNode entry : mosaics) {
			final JsonNode id = entry.get("mosaic").get("id");
			System.out.printf("  %s:%s%n",
				id.get("namespaceId").asText(),
				id.get("name").asText());
		}
		// [<step-4]
	}
}
